PerformanceTuning.md

```diff
@@ -0,0 +1,157 @@
+        }
+  ```
+
+* Disable Replicas (No Elasticsearch replica shard)
+ ```
+    Having replication set to 0, will avoid duplicate construction of graphs in 
+    both primary and replicas. When we enable replicas after the indexing, the 
+    serialized graphs are directly copied. Having no replicas means that losing 
+    a node(s) may incur data loss, so it is important that the data lives elsewhere 
+    so that this initial load can be retried in case of an issue.
+ ```
+More details [here](https://www.elastic.co/guide/en/elasticsearch/reference/master/tune-for-indexing-speed.html#_disable_replicas_for_initial_loads)
+    
+* Increase number of indexing threads
+  ```
+    If the hardware we choose has multiple cores, we could allow multiple threads 
+    in graph construction and there by speed up the indexing process. You could determine
+    the number of threads to be alloted by using the following setting   
+    https://github.com/opendistro-for-elasticsearch/k-NN#knnalgo_paramindex_thread_qty.
+     
+    Please keep an eye on CPU utilization and choose right number of threads. Since graph
+    construction is costly, having multiple threads can put additional load on CPU. 
+  ```
+    
+* Index all docs (Perform bulk indexing)
+
+* Forcemerge 
+  
+ Forcemerge is a costly operation and could take a while depending on number of segments and size of the segments.
+ To ensure force merge is completed, we could keep calling forcemerge with 5 minute interval till you get 200 response.
+    
+    curl -X POST "localhost:9200/myindex/_forcemerge?max_num_segments=1&pretty"
+    
+* Call refresh 
+
+ Calling refresh ensure the buffer is cleared and all segments are created so that documents are available for search. 
+ ```
+  POST /twitter/_refresh
+```
+* Add replicas (replica shards)
+ 
+ This will make replica shards come up with the already serialized graphs created on the primary shards during indexing. This way 
+ we avoid duplicate graph construction.
+
+* We can now enable replicas to copy the serialized graphs
+
+*  Enable refresh interval
+ ```
+      PUT /<index_name>/_settings
+        {
+            "index" : {
+                "refresh_interval" : "1m"
+            }
+        }
+ ```
+
+Please refer following doc (https://www.elastic.co/guide/en/elasticsearch/reference/master/tune-for-indexing-speed.html) for more details on improving indexing performance in general.
+
+##Search Performance Tuning
+
+To improve Search performance it is necessary to keep the number of segments under control. Lucene's IndexSearcher will search over all of the segments in a shard to find the 'size' best results. But, because the complexity of search for the HNSW algorithm is logarithmic with respect to the number of vectors, searching over 5 graphs with a 100 vectors each and then taking the top size results from ```5*k``` results will take longer than searching over 1 graph with 500 vectors and then taking the top size results from k results. 
```

Comment:
This needs a header like the other sections. Also, good explanation. I think adding a link to this newly created section from the force merge section would be helpful.

Reply:
makes sense